### PR TITLE
Fix stream usage in `segmented_gather()`

### PR DESCRIPTION
**Pull request by mythrocks**
`detail::segmented_gather()` inadvertently uses `cuda_default_stream` in some parts of its implementation, while using the user-specified stream in others.

This applies to the calls to `copy_range_in_place()`, `allocate_like()`, and `make_lists_column()`. ~This might produce race conditions, which might explain NVIDIA/spark-rapids/issues/4060. It's a rare failure that's quite hard to reproduce.~ This might lead to over-synchronization, though bad output is unlikely.

The commit here should sort this out, by switching to the `detail` APIs corresponding to the calls above.

### CODE REVIEW

**Comment by mythrocks** [comment]
Apologies. I realize it's late to include this in `21.12`. I can move this to `22.02` if it's preferable.

File: cpp/src/lists/copying/segmented_gather.cu [hunk 1]
```diff
@@ -13,6 +13,7 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
+#include <cudf/detail/copy_range.cuh>
 #include <cudf/detail/gather.cuh>
 #include <cudf/detail/gather.hpp>
```
Comment (ttnghia) [inline cpp/src/lists/copying/segmented_gather.cu:18]: You may want to remove one, since we may not need both at the same time. Please check.

**Review by hyperbolic2346 — APPROVED** [review]

**Comment by hyperbolic2346** [comment]
Thank you for doing this work and finding this issue.

File: cpp/src/lists/copying/segmented_gather.cu (hunk 1 above)
Comment (mythrocks) [reply cpp/src/lists/copying/segmented_gather.cu]: I've removed `<cudf/detail/gather.hpp>`, as per your suggestion.

**Review by ttnghia — APPROVED** [review]

**Comment by harrism** [comment]
Using the default stream should result in oversynchronization with non-default streams, not undersynchronization.

**Comment by mythrocks** [comment]
I'll have to keep hunting for the list extraction problem. 
In the meantime, maybe this PR should be redirected to the 22.02 release.

**Review by karthikeyann — APPROVED** [review]

**Comment by mythrocks** [comment]
Rebased for `branch-22.02`.

**Comment by mythrocks** [comment]
Rerun tests

**Comment by mythrocks** [comment]
Rerun tests

**Comment by mythrocks** [comment]
@gpucibot merge

**Comment by mythrocks** [comment]
This change has now been merged. The description was modified to indicate that corrupted output is unlikely.
Thank you for the reviews, all.

**Pull request merged**